mmedit/models/inpaintors/aot_inpaintor.py

```diff
@@ -0,0 +1,253 @@
+            # 0.5 for average the fake and real data
+            loss.update(loss_disc_shift=loss_d_shift * 0.5)
+
+        return loss
+
+    def generator_loss(self, fake_res, fake_img, data_batch):
+        """Forward function in generator training step.
+
+        In this function, we mainly compute the loss items for generator with
+        the given (fake_res, fake_img). In general, the `fake_res` is the
+        direct output of the generator and the `fake_img` is the composition of
+        direct output and ground-truth image.
+
+        Args:
+            fake_res (torch.Tensor): Direct output of the generator.
+            fake_img (torch.Tensor): Composition of `fake_res` and
+                ground-truth image.
+            data_batch (dict): Contain other elements for computing losses.
+
+        Returns:
+            tuple(dict): Dict contains the results computed within this
+                function for visualization and dict contains the loss items
+                computed in this function.
+        """
+        gt = data_batch['gt_img']
+        mask = data_batch['mask']
+        masked_img = data_batch['masked_img']
+
+        loss = dict()
+
+        if self.with_gan:
+            pred = self.disc(fake_img)
+            loss_g_fake = self.loss_gan(pred, True, False, mask=mask)
+            loss['loss_g_fake'] = loss_g_fake
+
+        if self.with_l1_valid_loss:
+            loss_loss_l1_valid = self.loss_l1_valid(fake_res, gt)
+            loss['loss_l1_valid'] = loss_loss_l1_valid
+
+        if self.with_out_percep_loss:
+            loss_out_percep, loss_out_style = self.loss_percep(fake_res, gt)
+            if loss_out_percep is not None:
+                loss['loss_out_percep'] = loss_out_percep
+            if loss_out_style is not None:
+                loss['loss_out_style'] = loss_out_style
+
+        res = dict(
+            gt_img=gt.cpu(),
+            masked_img=masked_img.cpu(),
+            fake_res=fake_res.cpu(),
+            fake_img=fake_img.cpu())
+
+        return res, loss
+
+    def forward_test(self,
+                     masked_img,
+                     mask,
+                     save_image=False,
+                     save_path=None,
+                     iteration=None,
+                     **kwargs):
+        """Forward function for testing.
+
+        Args:
+            masked_img (torch.Tensor): Tensor with shape of (n, 3, h, w).
+            mask (torch.Tensor): Tensor with shape of (n, 1, h, w).
+            save_image (bool, optional): If True, results will be saved as
+                image. Defaults to False.
+            save_path (str, optional): If given a valid str, the reuslts will
+                be saved in this path. Defaults to None.
+            iteration (int, optional): Iteration number. Defaults to None.
+
+        Returns:
+            dict: Contain output results and eval metrics (if exist).
+        """
+
+        masked_img = masked_img.float() + mask
```

Comment:
If I am not mistaken, the only difference to `forward_test` in `OneStageInpaintor` is this line. 

I think we can do the following instead of copying the entire function:
```
masked_img = masked_img.float() + mask
super().forward_test(masked_img, mask, save_image, save_path, iteration, **kwargs)

```

Reply:
Or could we modify the data pipeline instead? Then `forward_test` and `train_step` here can follow `OneStageInpaintor`.

Reply:
In the official code, the ssim and psnr metric are not weighted by the mask, so I also set `mask=None` in `forward_test` in aot_inpaintor here:

![Screenshot from 2022-01-20 10-06-11](https://user-images.githubusercontent.com/33197366/150252812-30459cd8-9132-4280-a085-f01e014207b8.png)

whereas in `OneStageInpaintor`, `mask` is not set to None when calculating psnr and ssim in `forward_test`

![Screenshot from 2022-01-20 10-08-32](https://user-images.githubusercontent.com/33197366/150253607-e7947563-2e7e-479e-8285-57bd6b714e7b.png)

Reply:
Okay then we can keep it as it is. But is it common that the holes are filled with different values? As you mentioned that some methods fill the holes with 1 and some fill with 0. It may be a good features to support such a choices when getting `masked_img` in the data pipeline.

Reply:
emm im not sure.. I think its only the AOT-GAN method that fills the holes with ones using `masked_img = masked_img.float() + mask` (maybe I am wrong, need to double check that). We may be able to support such a choice by adding an additional `fill_ones` parameter in `GetMaskedImage` in `mmedit.datasets.pipelines.formating.py`?

Reply:
Yes I think so. Instead of using `fill_ones`, I think it is better for users to specify the mask value (within 0 and 1). Then 1 - mask will represent valid regions.

Are you able to help with this?

Reply:
Is it necessary though, since most masks are simply binary mask?

Reply:
I was thinking about this too. But since fill_ones is a special case of supporting arbitrary values and it does not complicate the code much, I think it may be better to support a more general use case, in case some future methods have something to do with the mask (but it may eventually be useless haha)

Reply:
i see! But are we specifying the value of the hole region or the value of the mask itself? The latter is usually loaded from some files.

Reply:
Then that would complicate the situation a bit.

i am currently away from home, I will think about it and get back to you soon. Thanks!

Reply:
No worries! Thanks kelvin!

Reply:
@ckkelvinchan any update on this?

Reply:
Sorry, I forgot to reply to this thread. 

I think in this case we can use a simple argument `is_invert_mask`. If True, then we will swap the value of 1 and 0. Just forget the other values for now~

Reply:
emmm interesting. In our case the mask is not inverted (i.e. the masked region is still ones) but that `masked_img` is added to `mask` so as to fill the holes with ones. Wonder if `is_invert_mask` would do the trick here

Reply:
Okay I know what you mean. Then let's forget the `is_invert_mask`, and keep it as it is.